internal/controller/impersonatorconfig/impersonator_config.go

```diff
@@ -521,37 +524,109 @@ func (c *impersonatorConfigController) ensureClusterIPServiceIsStopped(ctx conte
+	updatedService.Spec.Selector = desiredService.Spec.Selector
+
+	// Do not simply overwrite the existing annotations with the desired annotations. Instead, merge-overwrite.
+	// Another actor in the system, like a human user or a non-Pinniped controller, might have updated the
+	// existing Service's annotations. If they did, then we do not want to overwrite those keys expect for
+	// the specific keys that are from the CredentialIssuer's spec, because if we overwrite keys belonging
+	// to another controller then we could end up infinitely flapping back and forth with the other controller,
+	// both updating that annotation on the Service.
+	if updatedService.Annotations == nil {
+		updatedService.Annotations = map[string]string{}
+	}
+	for k, v := range desiredService.Annotations {
+		updatedService.Annotations[k] = v
+	}
+
+	// Check if the the existing Service contains a record of previous annotations that were added by this controller.
+	// Note that in an upgrade, older versions of Pinniped might have created the Service without this bookkeeping annotation.
+	oldDesiredAnnotationKeysJSON, foundOldDesiredAnnotationKeysJSON := existingService.Annotations[annotationKeysKey]
+	oldDesiredAnnotationKeys := []string{}
+	if foundOldDesiredAnnotationKeysJSON {
+		_ = json.Unmarshal([]byte(oldDesiredAnnotationKeysJSON), &oldDesiredAnnotationKeys)
+		// In the unlikely event that we cannot parse the value of our bookkeeping annotation, just act like it
+		// wasn't present and update it to the new value that it should have based on the current desired state.
+	}
+
+	// Check if any annotations which were previously in the CredentialIssuer spec are now gone from the spec,
+	// which means that those now-missing annotations should get deleted.
+	// Nested loops are not efficient here, but these lists of annotations should be small.
+	for _, oldKey := range oldDesiredAnnotationKeys {
+		if !stringSliceContains(desiredAnnotationKeys, oldKey) {
+			delete(updatedService.Annotations, oldKey)
+		}
+	}
+
+	// If no annotations were requested, then remove the special bookkeeping annotation which might be
+	// leftover from a previous update. During the next update, non-existence will be taken to mean
+	// that no annotations were previously requested by the CredentialIssuer spec.
+	if len(desiredAnnotationKeys) == 0 {
+		delete(updatedService.Annotations, annotationKeysKey)
+	}
 
 	// If our updates didn't change anything, we're done.
-	if equality.Semantic.DeepEqual(existing, updated) {
+	if equality.Semantic.DeepEqual(existingService, updatedService) {
 		return nil
 	}
 
 	// Otherwise apply the updates.
 	c.infoLog.Info("updating service for impersonation proxy")
-	_, err = c.k8sClient.CoreV1().Services(c.namespace).Update(ctx, updated, metav1.UpdateOptions{})
+	_, err = c.k8sClient.CoreV1().Services(c.namespace).Update(ctx, updatedService, metav1.UpdateOptions{})
 	return err
 }
 
+func stringSliceContains(haystack []string, needle string) bool {
```

Comment:
Do we already have a function like this somewhere?

Reply:
Probably several 😂 . I think it's fine to define locally like this, at least until Go gets generics support.

Reply:
Removed in f17f7c0c6a53f50e6933a796b0804eb119a671cf.